src/plugins/vis_type_vega/public/data_model/vega_parser.ts

```diff
@@ -323,6 +324,52 @@ The URL is an identifier only. OpenSearch Dashboards and your browser will never
         delete this.spec.autosize;
       }
     }
+
+    if (this._config?.signals) {
+      Object.entries(this._config?.signals).forEach(([markId, signals]: [string, any]) => {
+        const mark = this.getMarkWithStyle(this.spec.marks, markId);
+
+        if (mark) {
+          signals.forEach((signal: Signal) => {
+            signal.on?.forEach((eventObj) => {
+              // We are prepending mark name here so that the signals only listens to the events on
+              // the elements related to this mark
+              eventObj.events = `@${mark.name}:${eventObj.events}`;
+            });
+          });
+          this.spec.signals = (this.spec.signals || []).concat(signals);
+        }
+      });
+    }
+  }
+
+  /**
+   * This method recursively looks for a mark that includes the given style.
+   * Returns undefined if it doesn't find it.
+   */
+  getMarkWithStyle(marks: any[], style: string): any {
```

Comment:
Seems like some unit tests here would be nice to validate all the conditional branching.